ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/MultiTenancyPrivacyController.java

```diff
@@ -127,6 +141,34 @@ public long determineBesuNonce(
         privacyGroupId, enclavePublicKey, callParams, blockNumber);
   }
 
+  @Override
+  public Optional<String> buildAndSendAddPayload(
+      final PrivateTransaction privateTransaction, final String enclaveKey) {
+    verifyPrivateFromMatchesEnclavePublicKey(
+        privateTransaction.getPrivateFrom().toBase64String(), enclaveKey);
+    verifyPrivacyGroupContainsEnclavePublicKey(
+        privateTransaction.getPrivacyGroupId().get().toBase64String(), enclaveKey);
+    return privacyController.buildAndSendAddPayload(privateTransaction, enclaveKey);
+  }
+
+  @Override
+  public PrivacyGroup retrievePrivacyGroup(
+      final String privacyGroupId, final String enclavePublicKey) {
+    verifyPrivacyGroupContainsEnclavePublicKey(privacyGroupId, enclavePublicKey);
```

Comment:
this is a bit funny, because the verifyPrivacyGroupContains ... method does call privacyController.retrievePrivacyGroup :-)
I think we should do the pirvacyController.retrievePrivacyGroup() call first and then check if the enclavePublicKey is part of the group before we are returning it!

Reply:
fixed